### PR TITLE
CDN Deployments

### PR DESCRIPTION
**Pull request by mikeshultz**
Deploy shops to a CDN

### TODO

- [x] Deploy to GCP bucket
- [x] Configure balancer
- [x] Configure CloudCDN
- [x] Adjust DNS configuration as necessary
- [x] Make deployments async
- [x] Test Route53
- [ ] Test Cloudflare DNS configuration
- [x] Fix CNAME conflicts when DNS records moved to CDN from IPFS gateway
- [x] Adjust custom domains to work with the CDN deployments

Ref #599

### CODE REVIEW

**Comment by mikeshultz** [comment]
I still want to do more testing before merge and deploy, but think it's ready for review whenever you all get the time.

**Comment by mikeshultz** [comment]
Unit tests have not been updated.  Looking at how that might work now.

File: backend/logic/deploy/bucket/gcp.js
Comment (franckc) [inline backend/logic/deploy/bucket/gcp.js:86]: type: Storage

File: backend/logic/deploy/bucket/gcp.js
Comment (franckc) [inline backend/logic/deploy/bucket/gcp.js:118]: Could we upload files in parallel (maybe in batch so that we control the max number of parallel uploads)? Might be a premature optimization though. I don't have a good crasp of the bucket.upload latency.

File: backend/logic/deploy/bucket/index.js
Comment (franckc) [inline backend/logic/deploy/bucket/index.js:27]: nit: it seems it could upload to multiple buckets if more than 1 module is configured?

File: backend/logic/deploy/build.js
Comment (franckc) [inline backend/logic/deploy/build.js:41]: yeah. I think we should throw. this should never happen

File: backend/logic/deploy/dns.js
Comment (franckc) [inline backend/logic/deploy/dns.js:9]: This does not seem like the correct description of what the method does? copy/pasta?

File: backend/logic/deploy/records.js
Comment (franckc) [inline backend/logic/deploy/records.js:2]: typo

File: backend/logic/deploy/records.js
Comment (franckc) [inline backend/logic/deploy/records.js:52]: why does the status need to be passed? shouldn't a new deployment always get created as Pending? Like we would create the deployment row, then pass it's id to the deploymentProcessor.

File: backend/logic/deploy/records.js
Comment (franckc) [inline backend/logic/deploy/records.js:46]: Why do we need a uuid? Could we simply use shop_deployments.id?

File: backend/queues/deploymentProcessor.js
Comment (franckc) [inline backend/queues/deploymentProcessor.js:46]: Is it possible success would be true but error would be set? under which scenario?

File: backend/logic/deploy/bucket/gcp.js
Comment (franckc) [inline backend/logic/deploy/bucket/gcp.js:56]: should we include the networkId to avoid any possible collision between shop deployments on different network but having the same authToken?

**Review by franckc — APPROVED** [review]
Looks really good. Super clean code. Love the use of asserts and the great jsdoc.
A few questions inline otherwise LGTM.

We'll definitely want to roll this out with caution since that's a lot of changes. Happy to help testing on staging once this gets merged.

File: backend/logic/deploy/bucket/gcp.js
Comment (mikeshultz) [reply backend/logic/deploy/bucket/gcp.js:118]: I'm currently not sure if GCP would allow that, so I'd prefer to leave that optimization until later.

File: backend/logic/deploy/build.js
Comment (mikeshultz) [reply backend/logic/deploy/build.js:41]: This was originally in `utils/deployShop`.  Do you know why this was originally a warning?

Comment (mikeshultz) [reply backend/logic/deploy/build.js:41]: https://github.com/OriginProtocol/dshop/blob/master/backend/utils/deployShop.js#L163

File: backend/logic/deploy/records.js
Comment (mikeshultz) [reply backend/logic/deploy/records.js:52]: hypothetical future thinking, but really probably unnecessary.  It's doing no harm right now, but you're probably right.

File: backend/logic/deploy/records.js
Comment (mikeshultz) [reply backend/logic/deploy/records.js:46]: I chose uuid because it makes things a lot cleaner.  A UUID can be created before the deployment process starts without having to fiddle with deployment locking DB records.

Specifically, the deployment DB record is first created in `deploy()` function, which is called in a bull queue job.  The only alternative would be to create the deployment DB record ahead of time, which doesn't really make sense, IMO.  Leaves complexity in the job processor instead of the express handler.

File: backend/queues/deploymentProcessor.js
Comment (mikeshultz) [reply backend/queues/deploymentProcessor.js:46]: Not currently, no

File: backend/logic/deploy/bucket/gcp.js
Comment (mikeshultz) [reply backend/logic/deploy/bucket/gcp.js:56]: I hadn't considered that.  Do we allow any `authToken` collisions in an instance?  I don't think so.  But since we may have multiple instances(for us, staging/prod), `BUCKET_PREFIX` should probably at least reflect that.

Comment (mikeshultz) [reply backend/logic/deploy/bucket/gcp.js:56]: I added `NODE_ENV` into the prefixes.

File: backend/logic/deploy/build.js
Comment (franckc) [reply backend/logic/deploy/build.js:41]: yeah. the code was eating that exception but that did not make any sense to me so I added the warning + TODO.
can you maybe check the warning does not show up in the logs? And if not let's throw an exception.

File: backend/logic/deploy/records.js
Comment (franckc) [reply backend/logic/deploy/records.js:46]: Sounds good. I'll leave it up to you to decide.

File: backend/logic/deploy/bucket/gcp.js
Comment (franckc) [reply backend/logic/deploy/bucket/gcp.js:56]: We don't allow collision on an instance (we check the authToken does not already exist in the DB before using it for a new shop).

I believe NODE_ENV is set to "production" for both staging and prod. I'd be slightly inclined to use networkId.

Comment (mikeshultz) [reply backend/logic/deploy/bucket/gcp.js:56]: > I believe NODE_ENV is set to "production" for both staging and prod so I'd be inclined to use networkId.

Well, that makes this tricky.  I still think there's a good possibility in the future for an individual node to support multiple networks, so I'm hesitant to use networkId.  Though... we're depending on networkConfig for credentials right now.  Maybe that is a good choice.

File: backend/logic/deploy/bucket/gcp.js
Comment (nick) [reply backend/logic/deploy/bucket/gcp.js:137]: I found a [code sample](https://github.com/googleapis/nodejs-storage/blob/master/samples/uploadDirectory.js#L69) that uses `Promise.all` to do it... but can optimize later

File: backend/logic/deploy/build.js
Comment (mikeshultz) [reply backend/logic/deploy/build.js:41]: There doesn't appear to have been any occurrences of this in the logs, at least within the last 6 months.  Will remove the catch here.

Comment (mikeshultz) [reply backend/logic/deploy/build.js:41]: Actually, just going to throw a more descriptive error.

**Review by nick — APPROVED** [review]
I read through the code and it looks good. Will try to run it locally too...

**Comment by mikeshultz** [comment]
Added some items to the TODO

**Comment by mikeshultz** [comment]
updated tests in https://github.com/OriginProtocol/dshop/pull/621/commits/ef82940137a27bedc945550c99bf2ee381254846.  Not super excited about the implementation of function overrides but it'll do.  Feedback welcome.

**Pull request merged**
